components/raftstore-v2/src/operation/command/admin/merge/commit.rs

```diff
@@ -0,0 +1,753 @@
+            source_region.get_region_epoch().get_version(),
+            region.get_region_epoch().get_version(),
+        ) + 1;
+        region.mut_region_epoch().set_version(version);
+        if keys::enc_end_key(&region) == keys::enc_start_key(source_region) {
+            region.set_end_key(source_region.get_end_key().to_vec());
+        } else {
+            region.set_start_key(source_region.get_start_key().to_vec());
+        }
+
+        // TODO: check both are trimmed.
+
+        let path = reg.tablet_path(self.region_id(), index);
+        // The last merge is incomplete.
+        if path.exists() {
+            std::fs::remove_dir_all(&path).unwrap();
+        }
+        let mut ctx = TabletContext::new(&region, Some(index));
+        ctx.flush_state = Some(self.flush_state().clone());
+        let tablet = reg.tablet_factory().open_tablet(ctx, &path).unwrap();
+        tablet.merge(&[&source_tablet, self.tablet()]).unwrap();
+        // This is only to consume in-memory tombstones, so that the two tablets don't
+        // overlap with each other. Remove this once trim check is finished.
+        tablet.flush_cfs(&[], true).unwrap();
+        self.set_tablet(tablet.clone());
+
+        self.region_state_mut().set_region(region.clone());
+        self.region_state_mut().set_state(PeerState::Normal);
+        self.region_state_mut().clear_merge_state();
+        self.region_state_mut().set_tablet_index(index);
+
+        PEER_ADMIN_CMD_COUNTER.commit_merge.success.inc();
+
+        Ok((
+            AdminResponse::default(),
+            AdminCmdResult::CommitMerge(CommitMergeResult {
+                index,
+                source_path,
+                region,
+                source: source_region.to_owned(),
+                tablet: Box::new(tablet),
+            }),
+        ))
+    }
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    // Called on source peer.
+    pub fn update_merge_progress_on_apply_res_prepare_merge<T: Transport>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+    ) {
+        assert!(self.applied_merge_state().is_some());
+        if let Some(c) = &self.merge_context().unwrap().catch_up_logs
+            && self.catch_up_logs_ready(c)
+        {
+            let c = self.merge_context_mut().catch_up_logs.take().unwrap();
+            self.apply_scheduler()
+                .unwrap()
+                .send(ApplyTask::LogsUpToDate(c));
+            return;
+        }
+        self.on_check_merge(store_ctx);
+    }
+
+    // Match v1::on_catch_up_logs_for_merge. Called on source peer.
+    pub fn on_catch_up_logs<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        mut catch_up_logs: CatchUpLogs,
+    ) {
+        let source_id = catch_up_logs.merge.get_source().get_id();
+        let target_id = catch_up_logs.target_region_id;
+        if source_id != self.region_id() {
+            if target_id != self.region_id() {
```

Comment:
I suggest to use different message type to make it more readable. I'm not a fan of v1 version.